pkg/agent/controller/bgp/controller.go

```diff
@@ -571,56 +588,95 @@ func (c *Controller) addServiceRoutes(advertisement *v1alpha1.ServiceAdvertiseme
 			if externalLocal && hasLocalEndpoints || !externalLocal {
 				for _, externalIP := range svc.Spec.ExternalIPs {
-					serviceIPs = append(serviceIPs, externalIP)
+					if c.enabledIPv4 && utilnet.IsIPv4String(externalIP) {
+						allRoutes[bgp.Route{Prefix: externalIP + ipv4Suffix}] = RouteMetadata{
+							Type:      ServiceExternalIP,
+							K8sObjRef: svc.GetNamespace() + "/" + svc.GetName(),
+						}
+					}
+					if c.enabledIPv6 && utilnet.IsIPv6String(externalIP) {
+						allRoutes[bgp.Route{Prefix: externalIP + ipv6Suffix}] = RouteMetadata{
+							Type:      ServiceExternalIP,
+							K8sObjRef: svc.GetNamespace() + "/" + svc.GetName(),
+						}
+					}
 				}
 			}
 		}
 		if ipTypes.Has(v1alpha1.ServiceIPTypeLoadBalancerIP) && svc.Spec.Type == corev1.ServiceTypeLoadBalancer {
 			if externalLocal && hasLocalEndpoints || !externalLocal {
-				serviceIPs = append(serviceIPs, getIngressIPs(svc)...)
+				loadBalancerIPs := getIngressIPs(svc)
+				for _, loadBalancerIP := range loadBalancerIPs {
+					if c.enabledIPv4 && utilnet.IsIPv4String(loadBalancerIP) {
+						allRoutes[bgp.Route{Prefix: loadBalancerIP + ipv4Suffix}] = RouteMetadata{
+							Type:      ServiceLoadBalancerIP,
+							K8sObjRef: svc.GetNamespace() + "/" + svc.GetName(),
+						}
+					}
+					if c.enabledIPv6 && utilnet.IsIPv6String(loadBalancerIP) {
+						allRoutes[bgp.Route{Prefix: loadBalancerIP + ipv6Suffix}] = RouteMetadata{
+							Type:      ServiceLoadBalancerIP,
+							K8sObjRef: svc.GetNamespace() + "/" + svc.GetName(),
+						}
+					}
+				}
 			}
 		}
 	}
-
-	for _, ip := range serviceIPs {
-		if c.enabledIPv4 && utilnet.IsIPv4String(ip) {
-			allRoutes.Insert(bgp.Route{Prefix: ip + ipv4Suffix})
-		}
-		if c.enabledIPv6 && utilnet.IsIPv6String(ip) {
-			allRoutes.Insert(bgp.Route{Prefix: ip + ipv6Suffix})
-		}
-	}
 }
 
-func (c *Controller) addEgressRoutes(allRoutes sets.Set[bgp.Route]) {
+func (c *Controller) addEgressRoutes(allRoutes map[bgp.Route]RouteMetadata) {
 	egresses, _ := c.egressLister.List(labels.Everything())
 	for _, eg := range egresses {
 		if eg.Status.EgressNode != c.nodeName {
 			continue
 		}
 		ip := eg.Status.EgressIP
 		if c.enabledIPv4 && utilnet.IsIPv4String(ip) {
-			allRoutes.Insert(bgp.Route{Prefix: ip + ipv4Suffix})
+			allRoutes[bgp.Route{Prefix: ip + ipv4Suffix}] = RouteMetadata{
+				Type:      EgressIP,
+				K8sObjRef: eg.Name,
+			}
 		}
 		if c.enabledIPv6 && utilnet.IsIPv6String(ip) {
-			allRoutes.Insert(bgp.Route{Prefix: ip + ipv6Suffix})
+			allRoutes[bgp.Route{Prefix: ip + ipv6Suffix}] = RouteMetadata{
+				Type:      EgressIP,
+				K8sObjRef: eg.Name,
+			}
 		}
 	}
 }
 
-func (c *Controller) addPodRoutes(allRoutes sets.Set[bgp.Route]) {
+func (c *Controller) addPodRoutes(allRoutes map[bgp.Route]RouteMetadata) {
 	if c.enabledIPv4 {
-		allRoutes.Insert(bgp.Route{Prefix: c.podIPv4CIDR})
+		allRoutes[bgp.Route{Prefix: c.podIPv4CIDR}] = RouteMetadata{
+			Type:      NodeIPAMPodCIDR,
+			K8sObjRef: "Pod",
```

Comment:
How about renaming it to `PodCIDR`?

Reply:
`K8sObjRef` holds the object name. Here we cannot provide a specific Pod name as the route is not PodIP but PodCIDR. The Type already clarifies that it is PodCIDR, so i don't  think we should have PodCIDR in K8sObjRef as well.